agg-ast/schema_derivation/src/lib.rs

```diff
@@ -576,3 +576,113 @@ macro_rules! array_element_schema_or_error {
+                            // technically this should be the same db as the current_db, but it is worth
+                            // handling in case we can deal with cross database lookups in the future.
+                            Some(agg_ast::definitions::LookupFrom::Namespace(ns)) => {
+                                (ns.database.clone(), Some(ns.collection.clone()))
+                            }
+                            _ => (current_db.clone(), None),
+                        };
+                        // we directly recurse on the subpipeline because the from collection will be
+                        // added to the set as the first step of the recursive call
+                        namespaces.append(&mut get_namespaces_for_pipeline(
+                            pipeline,
+                            from_db,
+                            from_collection,
+                        ));
+                    }
+                };
+            }
+            agg_ast::definitions::Stage::GraphLookup(graph_lookup) => {
+                add_namespace!(graph_lookup.from);
+            }
+            agg_ast::definitions::Stage::UnionWith(union_with) => match union_with {
+                agg_ast::definitions::UnionWith::Collection(c) => {
+                    namespaces.insert(agg_ast::definitions::Namespace::new(current_db.clone(), c));
+                }
+                agg_ast::definitions::UnionWith::Pipeline(union_with_pipepline) => {
+                    if let Some(pipeline) = union_with_pipepline.pipeline {
+                        namespaces.append(&mut get_namespaces_for_pipeline(
+                            pipeline,
+                            current_db.clone(),
+                            union_with_pipepline.coll,
+                        ));
+                    } else if let Some(c) = union_with_pipepline.coll.as_ref() {
+                        add_namespace!(c.clone());
+                    };
+                }
+            },
+            // collection, join, and equijoin are sql constructs that get converted into lookup stages; however, we
+            // handle here for the sake of completeness.
+            agg_ast::definitions::Stage::Join(join) => {
+                let db = join.database.unwrap_or_else(|| current_db.clone());
+                if let Some(coll) = join.collection {
+                    namespaces.insert(agg_ast::definitions::Namespace::new(db, coll));
+                }
+            }
+            agg_ast::definitions::Stage::EquiJoin(ej) => {
+                let db = ej.database.unwrap_or_else(|| current_db.clone());
+                if let Some(coll) = ej.collection {
+                    namespaces.insert(agg_ast::definitions::Namespace::new(db, coll));
+                }
+            }
+            agg_ast::definitions::Stage::Collection(c) => {
+                namespaces.insert(agg_ast::definitions::Namespace::new(c.db, c.collection));
+            }
+            _ => {}
```

Comment:
It looks like Facet was missed?

```suggestion
            agg_ast::definitions::Stage::Facet(f) => {
                for (_, stages) in f {
                    namespaces.append(&mut get_namespaces_for_pipeline(
                        stages,
                        current_db.clone(),
                        current_collection.clone(),
                    ));
                }
            }
            _ => {}
```